python/lsst/daf/butler/direct_query_driver/_driver.py

```diff
@@ -0,0 +1,1200 @@
+        number of rows in a page may be smaller due to postprocessing.
+    constant_rows_limit : `int`, optional
+        Maximum number of uploaded rows to include in queries via
+        `Database.constant_rows`; above this limit a temporary table is used
+        instead.
+    postprocessing_filter_factor : `int`, optional
+        The number of database rows we expect to have to fetch to yield a
+        single output row for queries that involve postprocessing.  This is
+        purely a performance tuning parameter that attempts to balance between
+        fetching too much and requiring multiple fetches; the true value is
+        highly dependent on the actual query.
+    """
+
+    def __init__(
+        self,
+        db: Database,
+        universe: DimensionUniverse,
+        managers: RegistryManagerInstances,
+        defaults: RegistryDefaults,
+        raw_page_size: int = 10000,
+        constant_rows_limit: int = 1000,
+        postprocessing_filter_factor: int = 10,
+    ):
+        self.db = db
+        self.managers = managers
+        self._universe = universe
+        self._defaults = defaults
+        self._materializations: dict[
+            qt.MaterializationKey, tuple[sqlalchemy.Table, frozenset[str], Postprocessing]
+        ] = {}
+        self._upload_tables: dict[qt.DataCoordinateUploadKey, sqlalchemy.FromClause] = {}
+        self._exit_stack: ExitStack | None = None
+        self._raw_page_size = raw_page_size
+        self._postprocessing_filter_factor = postprocessing_filter_factor
+        self._constant_rows_limit = constant_rows_limit
+        self._cursors: dict[PageKey, _Cursor] = {}
+
+    def __enter__(self) -> None:
+        self._exit_stack = ExitStack()
+        # It might be nice to defer opening a transaction here until first use
+        # to reduce the time spent in transactions.  But it's worth noting that
+        # this is the default low-level behavior of the Python SQLite driver,
+        # and it makes it incredibly prone to deadlocks.  We might be okay
+        # here, because Query doesn't do true write operations - just temp
+        # table writes - but I'm not confident that's enough to make delayed
+        # transaction starts safe against deadlocks, and it'd be more
+        # complicated to implement anyway.
+        #
+        # We start a transaction rather than just opening a connection both
+        # to establish consistency guarantees throughout the query context and
+        # to make temp table and cursors work with pg_bouncer transaction
+        # affinity.
+        self._exit_stack.enter_context(self.db.transaction(for_temp_tables=True))
+
+    def __exit__(self, exc_type: Any, exc_value: Any, traceback: Any) -> None:
+        assert self._exit_stack is not None
+        while self._cursors:
+            _, cursor = self._cursors.popitem()
+            cursor.close(exc_type, exc_value, traceback)
```

Comment:
I was just thinking call `exit_stack.push(cursor)` right here instead of calling `cursor.close()`... `push()` calls `__exit__` without calling `__enter__`.